src/goto-symex/complexity_module.cpp

```diff
@@ -0,0 +1,205 @@
+#include "goto_symex_state.h"
+#include <cmath>
+
+complexity_modulet::complexity_modulet(
+  message_handlert &message_handler,
+  const optionst &options)
+  : log(message_handler)
+{
+  std::size_t limit = options.get_signed_int_option("symex-complexity-limit");
+  if((complexity_active = limit > 0))
+  {
+    // This gives a curve that allows low limits to be rightly restrictive,
+    // while larger numbers are very large.
+    max_complexity = static_cast<std::size_t>((limit * limit) * 25);
+
+    const std::size_t failed_children_limit =
+      options.get_signed_int_option("symex-complexity-failed-children-limit");
+    const std::size_t unwind = options.get_signed_int_option("unwind");
+
+    // In order of priority:
+    //  * explicit limit
+    //  * inferred limit from unwind
+    //  * best limit we can apply with very little information.
+    if(failed_children_limit > 0)
+      max_loops_complexity = failed_children_limit;
+    else if(unwind > 0)
+      max_loops_complexity = std::max(static_cast<int>(floor(unwind / 3)), 1);
+    else
+      max_loops_complexity = limit;
+  }
+}
+
+/// Takes into account nested loops and various other subtleties.
+optionalt<std::reference_wrapper<framet::active_loop_infot>>
+complexity_modulet::get_current_active_loop(call_stackt &current_call_stack)
+{
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    // As we're walking bottom-up, the first frame we find with active loops
+    // is our closest active one.
+    if(!frame_iter->active_loops.empty())
+    {
+      return std::reference_wrapper<framet::active_loop_infot>(
+        frame_iter->active_loops.back());
+    }
+  }
+
+  return {};
+}
+
+bool complexity_modulet::in_blacklisted_loop(
+  const call_stackt &current_call_stack,
+  const goto_programt::const_targett &instr)
+{
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    for(auto &loop_iter : frame_iter->active_loops)
+    {
+      for(auto &blacklisted_loop : loop_iter.blacklisted_loops)
+      {
+        if(blacklisted_loop.get().contains(instr))
+        {
+          return true;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+bool complexity_modulet::are_loop_children_too_complicated(
+  call_stackt &current_call_stack)
+{
+  // Check that current loops don't violate complexity limits, if they
+  // do then add that loop to the blacklist for all currently-running loops.
+
+  // We want to make sure that if a certain execution of a loop is too complex
+  // no matter the branching, that we cancel out early. This is critical because
+  // otherwise we will continue to execute nested loops that we already know
+  // are far too complicated to deal with, wasting huge amounts of time.
+  std::size_t sum_complexity = 0;
+
+  // If non-empty, is the lowest-level loop we should blacklist in the stack.
```

Comment:
Not clear what the comment means. How about `Walk from the innermost to outermost loop on the stack, blacklisting any loop (and all of its parents) that are too complex`?